### ISSUE TITLE
Limiting additional memory pools

### ISSUE THREAD

**Issue by ipsi** [cloudfoundry/java-buildpack#319]
@nebhale This is the first issue we talked about the other day.

Currently, the Java Buildpack & Memory Calculator look at the following memory spaces:
- Heap
- Meta/PermGen Space
- Stack Size
- "Native"

Where Native includes an estimation of the amount used by NIO direct memory, the code cache, the compressed class space, the (tiny) amount used by initd, bash, memory mapped files (e.g. libjvm.so), etc. The ones that Java gives us tools to limit are the NIO direct memory, the code cache, and the compressed class space.

I think it would be sensible to limit those extra pools that we can, as, by default, their bounds are _extremely_ high, and should they fill up (unlikely, but theoretically possible), the application will crash with exit code 255 (I think), and no detailed information about what caused the crash (I believe Garden are looking into providing better error messages?).

The flags that would be necessary for this are:
- `-XX:InitialCodeCacheSize` (Default: unsure)
- `-XX:ReservedCodeCacheSize` (Default: 240MiB; Will never throw an OOM, but code compilation will halt if no free memory - performance implications)
- `-XX:CompressedClassSpaceSize` (Default: 1GiB)
- `-XX:MaxDirectMemorySize` (Default: unbounded)

The amount that will actually be used for each are broadly dependent on the following:
- `-XX:InitialCodeCacheSize`, `-XX:ReservedCodeCacheSize`, `-XX:CompressedClassSpaceSize`: Number of classes in use - the more classes there are, the more _things_ go into the compressed class space, and the more code could be compiled into the code cache.
- `-XX:MaxDirectMemorySize`: Number of NIO `ByteBuffer`s (and etc) used. Unless an application makes direct use of them, only Tomcat (or another container) will be using `ByteBuffer`s (AFAIK), so the maximum amount of space used would be porportional to the average response size and the number of threads.

I think that limiting these pools is a good idea, because if Java throws an OutOfMemory error, that makes it a lot easier to find the cause, rather than relying on Garden to provide us enough information to find the root cause.

**Comment by ipsi** [comment]
On a semi-related subject, I've been wondering if the type of application deployed to CF is specific enough that the buildpack should override a number of the default garbage collection parameters. Maybe this is something you've already been thinking about / have thought about at great length, but if not:

The idea is that applications deployed to CF should obey the 12 factors.

Given that, the following assumptions are probably reasonable for a "typical" CF application:

* Large number of concurrent threads
* Large number of concurrent, short-lived objects created
* Long-lived objects allocated almost exclusively on start up
* Long-lived objects allocated after start up likely to be a case of rat-packing, rather than genuine need for tenure.

I think the default settings for Java assume that it's going to be a long-running process *which has to remember a lot of state*, but that's not the case here. Indeed, I'd say that the ideal CF application (whether that's "most", "some", or "practically none" is a different story...) has a very particular memory allocation model. Combine that with the intent of CF to make scaling easy, and it seems like would make sense to tune the garbage collector for the above scenario. Meaning:

* Look at using G1 or CMS, which focus on latency.
  * If throughput is a concern, add more instances
* Optimise GC for a large number of short-lived objects
  * This would probably involve setting a large Young generation size, and possibly increasing the number of generations before tenure.

Given how big an impact GC can have on performance, I think it's worth looking at whether the type of application deployed to CF would benefit from changing the default GC options.

Unfortunately, tuning an application for optimal GC is very application-specific. Perhaps it would make sense to provide some high-level options to enable logging of GC? Maybe even writing some tools to analyse the output... But that's a very different rabbit hole to start going down.

Some references:

https://www.infoq.com/articles/tuning-tips-G1-GC

http://product.hubspot.com/blog/g1gc-fundamentals-lessons-from-taming-garbage-collection (quite recent, micro-service oriented)

https://engineering.linkedin.com/garbage-collection/garbage-collection-optimization-high-throughput-and-low-latency-java-applications (a bit old)

**Comment by nebhale** [comment]
@ipsi Yeah, it's a reasonable thing to look into, but at this point the number of people who aren't even doing 12-Factor but rather lifting and shifting into CF means that we can't really make these optimizations.  Instead we have to kind of target lowest common denominator kinds of optimizations and allow users to do more sophisticated changes if they can.  In addition, we've got some back and forth with [Skipjaq](http://www.skipjaq.com) to see if they can automate these kind of performance optimizations without requiring explicit configuration.

**Comment by TimGerlach** [comment]
@ipsi Nice to see that there are more and more people interested in improving the memory calculation! I totally agree with your thoughts.

We faced similar issues with random app crashes due to exceeded container memory quota like:

```
time                          event         actor         description
2015-03-06T08:40:51.00-0500   app.crash     my-app-name   index: 0, reason: CRASHED, exit_description: out of memory, exit_status: 255
```

We looked into that topic and are currently thinking about how to improve the memory calculation in order to prevent (or at least reduce) such crashes in the future. In my opinion all memory related crashes (of Java apps) should lead to a ``java.lang.OutOfMemoryError`` with detailed descriptions rather than containers being killed by the OOM killer.

I found a lot of related issues created by @lhotari (thanks for that) and we tried to extend the memory forecasting @lhotari mentioned in #157 as:

```
Max memory = [-Xmx] + [-XX:MaxMetaspaceSize] + [-XX:ReservedCodeCacheSize] + [-XX:CompressedClassSpaceSize] + [XX:MaxDirectMemorySize] + ([-Xss] * maximum_thread_count)
```

I think it makes totally sense to limit the ``native`` areas explicitly by default. This includes, for example, limiting the CodeCache with ``-XX:ReservedCodeCacheSize=``. Additionally, one could think about setting expansion sizes explicitly. Both together would help in forecasting the memory consumption of the Java App and reduce memory fragmentation as discussed in #157.

**Comment by glyn** [comment]
@ipsi There are some changes to the memory calculator in progress ([1]) which will default `-XX:MaxDirectMemorySize` to a fixed value and `-XX:CompressedClassSpaceSize`, `-XX:MaxMetaspaceSize`, and `-XX:ReservedCodeCacheSize` to a linear function of classes the application is likely to load (which the Java buildpack will again default to be another linear function of the number of .class & .groovy files in the application, including any nested JARs). The user can override any of these simply by setting a corresponding value in `JAVA_OPTS`.

So this leaves `-XX:InitialCodeCacheSize` from the list in the description of this issue. The memory calculator is currently steering away from defaulting minimum sizes and so I proposed not to default `-XX:InitialCodeCacheSize` until there is evidence that this would be generally beneficial. We want to avoid over-complicating the memory calculator given that we ended up with something that was quite hard for the users to understand and use last time.

[1] https://www.pivotaltracker.com/story/show/135936599
